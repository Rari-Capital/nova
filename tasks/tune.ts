import chalk from "chalk";
import ora from "ora";
import chai from "chai";
chai.should();

import { task } from "hardhat/config";
import { TransactionResponse } from "@ethersproject/abstract-provider";

import { L1NovaExecutionManager } from "../typechain";

task("tune", "Tunes an execution manager's missing gas estimate based on single exec tx.")
  .addParam("txHash", "Transaction hash of an exec tx.")
  .addFlag(
    "update",
    "If enabled, the task will update the estimate on-chain. If not enabled the task will just report the optimal estimate."
  )
  .setAction(async ({ txHash, update }, hre) => {
    const txPromise = hre.ethers.provider.getTransaction(txHash);
    const resolvedTx = await txPromise;

    const L1_NovaExecutionManager: L1NovaExecutionManager = (
      await hre.ethers.getContractFactory("L1_NovaExecutionManager")
    ).attach(resolvedTx.to) as any;

    if (update) {
      await tuneMissingGasEstimate(L1_NovaExecutionManager, txPromise);
    } else {
      console.log();

      const loader = ora({
        text: chalk.gray(`finding optimal missing gas estimate for tx: ${chalk.yellow(txHash)}\n`),
        color: "yellow",
        indent: 6,
      }).start();

      const { currentMissingGasEstimate, optimalMissingGasEstimate } =
        await findOptimalMissingGasEstimate(L1_NovaExecutionManager, txPromise);

      loader.stopAndPersist({
        symbol: chalk.yellow("✓"),
        text: chalk.gray(
          `based on this tx-hash, the optimal missing gas estimate would be ${chalk.yellow(
            optimalMissingGasEstimate.toString()
          )} (currently ${chalk.yellow(currentMissingGasEstimate.toString())})\n`
        ),
      });

      loader.indent = 0;
    }
  });

/**
 * Finds the optimal missing gas estimate for an execution manager based on a single exec tx.
 */
async function findOptimalMissingGasEstimate(
  L1_NovaExecutionManager: L1NovaExecutionManager,
  tx: Promise<TransactionResponse>
) {
  const { gasUsed, logs } = await (await tx).wait();

  // Since we accept a generic TransactionResponse we have to parse the logs manually.
  const execEvent = L1_NovaExecutionManager.interface.parseLog(logs[logs.length - 1]);

  // We are assuming that the current gas config was the same when the
  // tx was executed. Don't try to use tune on really old transactions!
  const currentGasConfig = await L1_NovaExecutionManager.gasConfig();

  const currentMissingGasEstimate = currentGasConfig.missingGasEstimate.toNumber();
  const underestimateAmount = gasUsed.toNumber() - execEvent.args.gasUsed.toNumber();

  return {
    currentGasConfig,
    currentMissingGasEstimate,
    optimalMissingGasEstimate: currentMissingGasEstimate + underestimateAmount,
  };
}

/**
 * Tunes an execution manager's missing gas estimate based on single exec tx.
 */
export async function tuneMissingGasEstimate(
  L1_NovaExecutionManager: L1NovaExecutionManager,
  tx: Promise<TransactionResponse>
) {
  const { currentGasConfig, currentMissingGasEstimate, optimalMissingGasEstimate } =
    await findOptimalMissingGasEstimate(L1_NovaExecutionManager, tx);

  console.log();

  const loader = ora({
    text: chalk.gray(
      `tuning missing gas estimate from ${chalk.magenta(
        currentMissingGasEstimate.toString()
      )} to ${chalk.magenta(optimalMissingGasEstimate.toString())}\n`
    ),
    color: "magenta",
    indent: 6,
  }).start();

  // Add a 500 gas buffer for a bit of leeway.
  const newMissingGasEstimate = optimalMissingGasEstimate + 500;

  // Update the missing gas estimate on-chain.
  await (
    await L1_NovaExecutionManager.updateGasConfig({
      ...currentGasConfig,
      missingGasEstimate: newMissingGasEstimate,
    })
  ).wait();

  loader.stopAndPersist({
    symbol: chalk.magenta("✓"),
    text: chalk.gray(
      `tuned missing gas estimate from ${chalk.magenta(
        currentMissingGasEstimate.toString()
      )} to ${chalk.magenta(optimalMissingGasEstimate.toString())}\n`
    ),
  });

  loader.indent = 0;
}
