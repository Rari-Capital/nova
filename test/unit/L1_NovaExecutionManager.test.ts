import { ethers } from "hardhat";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";

import {
  checkAllFunctionsForAuth,
  deployStrategy,
  executeRequest,
  getFactory,
  snapshotGasCost,
  StrategyRiskLevel,
} from "../../utils/testUtils";
import { tuneMissingGasEstimate } from "../../tasks/tune";

import {
  L1NovaExecutionManager,
  L1NovaExecutionManager__factory,
  MockStrategy,
  MockCrossDomainMessenger,
  MockCrossDomainMessenger__factory,
  MockERC20,
  MockERC20__factory,
  NoReturnValueERC20__factory,
  BadReturnValueERC20__factory,
  ReturnFalseERC20__factory,
  MockAuthority__factory,
} from "../../typechain";

describe("L1_NovaExecutionManager", function () {
  let signers: SignerWithAddress[];
  before(async () => {
    signers = await ethers.getSigners();
  });

  // Nova Contracts:
  let L1_NovaExecutionManager: L1NovaExecutionManager;

  /// Mocks:
  let MockERC20: MockERC20;
  let MockCrossDomainMessenger: MockCrossDomainMessenger;

  // Strategies:
  let UnknownStrategy: MockStrategy;
  let SafeStrategy: MockStrategy;
  let UnsafeStrategy: MockStrategy;

  describe("constructor/setup", function () {
    it("should properly deploy mocks", async function () {
      MockERC20 = await (await getFactory<MockERC20__factory>("MockERC20")).deploy();

      MockCrossDomainMessenger = await (
        await getFactory<MockCrossDomainMessenger__factory>("MockCrossDomainMessenger")
      ).deploy();
    });

    it("should properly deploy the execution manager", async function () {
      L1_NovaExecutionManager = await (
        await getFactory<L1NovaExecutionManager__factory>("L1_NovaExecutionManager")
      ).deploy(ethers.constants.AddressZero, MockCrossDomainMessenger.address);
    });

    it("should not allow calling authed functions before permitted", async function () {
      const [, nonDeployer] = signers;

      await checkAllFunctionsForAuth(L1_NovaExecutionManager, nonDeployer);
    });

    it("should allow changing the execution manager's authority", async function () {
      const MockAuthority = await (
        await getFactory<MockAuthority__factory>("MockAuthority")
      ).deploy();

      // Set the authority to a MockAuthority that always returns true.
      await L1_NovaExecutionManager.setAuthority(MockAuthority.address);
    });

    it("should properly deploy strategies", async function () {
      UnknownStrategy = await deployStrategy(L1_NovaExecutionManager, StrategyRiskLevel.UNKNOWN);

      SafeStrategy = await deployStrategy(L1_NovaExecutionManager, StrategyRiskLevel.SAFE);

      UnsafeStrategy = await deployStrategy(L1_NovaExecutionManager, StrategyRiskLevel.UNSAFE);
    });

    it("should properly use constructor arguments", async function () {
      // Make sure the constructor params were properly entered.
      await L1_NovaExecutionManager.CROSS_DOMAIN_MESSENGER().should.eventually.equal(
        MockCrossDomainMessenger.address
      );
      await L1_NovaExecutionManager.L2_NOVA_REGISTRY_ADDRESS().should.eventually.equal(
        ethers.constants.AddressZero
      );
    });

    it("should contain constants that match expected values", async function () {
      // Make sure the hard revert text is correct.
      await L1_NovaExecutionManager.HARD_REVERT_TEXT().should.eventually.equal(
        "__NOVA__HARD__REVERT__"
      );
    });

    it("should allow tuning the missing gas estimate", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        nonce: 420,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData("thisFunctionWillNotRevert"),

        // It will overestimate before tuning.
        expectedGasOverestimateAmount: 99999999999999,
      });

      await tuneMissingGasEstimate(L1_NovaExecutionManager, tx);
    });
  });

  describe("registerSelfAsStrategy", function () {
    it("should not allow registering as UNKNOWN", async function () {
      const strategy = await deployStrategy(L1_NovaExecutionManager);

      await strategy
        .registerSelfAsStrategy(StrategyRiskLevel.UNKNOWN)
        .should.be.revertedWith("INVALID_RISK_LEVEL");
    });

    it("should not allow registering multiple times", async function () {
      const strategy = await deployStrategy(L1_NovaExecutionManager);

      await strategy.registerSelfAsStrategy(StrategyRiskLevel.SAFE);

      await strategy
        .registerSelfAsStrategy(StrategyRiskLevel.UNSAFE)
        .should.be.revertedWith("ALREADY_REGISTERED");
    });

    it("should allow registering properly", async function () {
      const strategy = await deployStrategy(L1_NovaExecutionManager);

      await strategy.registerSelfAsStrategy(StrategyRiskLevel.UNSAFE);

      await L1_NovaExecutionManager.getStrategyRiskLevel(strategy.address).should.eventually.equal(
        StrategyRiskLevel.UNSAFE
      );
    });
  });

  describe("exec", function () {
    it("should revert if a hard revert is triggered", async function () {
      const [relayer] = signers;

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData("thisFunctionWillHardRevert"),
      }).should.be.revertedWith("HARD_REVERT");
    });

    it("respects the deadline", async function () {
      const [relayer] = signers;

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        // Set a deadline 60 seconds in the past
        deadline: Math.floor(Date.now() / 1000) - 60,
      }).should.be.revertedWith("PAST_DEADLINE");
    });

    it("should not allow calling the cross domain messenger", async function () {
      const [relayer] = signers;

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: MockCrossDomainMessenger.address,
        l1Calldata: "0x00",
      }).should.be.revertedWith("UNSAFE_STRATEGY");
    });

    it("should not allow calling the approval escrow", async function () {
      const [relayer] = signers;

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: await L1_NovaExecutionManager.L1_NOVA_APPROVAL_ESCROW(),
        l1Calldata: "0x00",
      }).should.be.revertedWith("UNSAFE_STRATEGY");
    });

    it("should not allow self calls", async function () {
      const [relayer] = signers;

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: L1_NovaExecutionManager.address,
      }).should.be.revertedWith("UNSAFE_STRATEGY");
    });

    it("should not allow reentrancy", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData("thisFunctionWillTryToReenter"),
      });

      await tx.should.emit(SafeStrategy, "ReentrancyFailed");
    });

    it("should properly execute a minimal exec", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData("thisFunctionWillNotRevert"),
      });
      SafeStrategy;

      await snapshotGasCost(tx);
    });

    it("should properly execute a stateful exec", async function () {
      const [relayer] = signers;

      await SafeStrategy.counter().should.eventually.equal(1);

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData("thisFunctionWillModifyState"),
      });

      await snapshotGasCost(tx);

      await SafeStrategy.counter().should.eventually.equal(2);
    });

    it("should not revert due to a soft revert", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData("thisFunctionWillRevert"),
        shouldSoftRevert: true,
      });

      await snapshotGasCost(tx);
    });

    it("should not revert due to a hard revert triggered by a SAFE strategy", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData("thisFunctionWillHardRevert"),
        shouldSoftRevert: true,
      });

      await snapshotGasCost(tx);
    });

    it("should not revert due to a hard revert triggered by an UNKNOWN strategy", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnknownStrategy.address,
        l1Calldata: UnknownStrategy.interface.encodeFunctionData("thisFunctionWillHardRevert"),
        shouldSoftRevert: true,
      });

      await snapshotGasCost(tx);
    });
  });

  describe("transferFromRelayer", function () {
    it("should transfer an arbitrary token to a strategy when requested", async function () {
      const [relayer] = signers;

      // Approve the right amount of input tokens.
      const weiAmount = ethers.utils.parseEther("1337");
      await MockERC20.approve(await L1_NovaExecutionManager.L1_NOVA_APPROVAL_ESCROW(), weiAmount);

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayer",
          [MockERC20.address, weiAmount]
        ),
        expectedGasOverestimateAmount: 4800, // 4800 more than expected will be consumed due to a gas refund.
      });

      await snapshotGasCost(tx)
        // The correct amount of tokens should be transferred to the strategy.
        .should.emit(MockERC20, "Transfer")
        .withArgs(relayer.address, UnsafeStrategy.address, weiAmount);

      await MockERC20.balanceOf(UnsafeStrategy.address).should.eventually.equal(weiAmount);
    });

    it("will properly handle a transferFrom with no return value", async function () {
      const [relayer] = signers;

      const NoReturnValueERC20 = await (
        await getFactory<NoReturnValueERC20__factory>("NoReturnValueERC20")
      ).deploy();

      // Since this is a mock token we don't have to approve this amount.
      const weiAmount = ethers.utils.parseEther("1337");

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayer",
          [NoReturnValueERC20.address, weiAmount]
        ),
      });

      // Should emit an event to indicate it was called properly.
      await snapshotGasCost(tx)
        .should.emit(NoReturnValueERC20, "Transfer")
        .withArgs(relayer.address, UnsafeStrategy.address, weiAmount);
    });

    it("will hard revert if tokens were not approved", async function () {
      const [relayer] = signers;

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayer",
          [MockERC20.address, ethers.utils.parseEther("9999999")]
        ),
      }).should.be.revertedWith("HARD_REVERT");
    });

    it("will hard revert if transferFrom returns a non-bool", async function () {
      const [relayer] = signers;

      const BadReturnValueERC20 = await (
        await getFactory<BadReturnValueERC20__factory>("BadReturnValueERC20")
      ).deploy();

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayer",
          [BadReturnValueERC20.address, 0]
        ),
      }).should.be.revertedWith("HARD_REVERT");
    });

    it("will hard revert if transferFrom returns false without reverting", async function () {
      const [relayer] = signers;

      const ReturnFalseERC20 = await (
        await getFactory<ReturnFalseERC20__factory>("ReturnFalseERC20")
      ).deploy();

      await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayer",
          [ReturnFalseERC20.address, 0]
        ),
      }).should.be.revertedWith("HARD_REVERT");
    });

    it("will not allow anyone to call if not executing", async function () {
      await L1_NovaExecutionManager.transferFromRelayer(
        MockERC20.address,
        0
      ).should.be.revertedWith("NOT_CURRENT_STRATEGY");
    });

    it("will not allow the previous strategy to call if not executing", async function () {
      await UnsafeStrategy.thisFunctionWillTransferFromRelayer(
        MockERC20.address,
        1
      ).should.be.revertedWith("NO_ACTIVE_EXECUTION");
    });

    it("will not allow a random contract to call during execution", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnsafeStrategy.address,
        l1Calldata: UnsafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillEmulateAMaliciousExternalContractTryingToStealRelayerTokens",
          [MockERC20.address, 0]
        ),
      });

      await snapshotGasCost(tx).should.emit(UnsafeStrategy, "StealRelayerTokensFailed");
    });

    it("will not allow a SAFE strategy to call", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: SafeStrategy.address,
        l1Calldata: SafeStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayerAndExpectUnsupportedRiskLevel",
          [MockERC20.address, 0]
        ),
      });

      await snapshotGasCost(tx).should.emit(
        SafeStrategy,
        "TransferFromRelayerFailedWithUnsupportedRiskLevel"
      );
    });

    it("will not allow an UNKNOWN strategy to call", async function () {
      const [relayer] = signers;

      const { tx } = await executeRequest(L1_NovaExecutionManager, {
        relayer: relayer.address,
        strategy: UnknownStrategy.address,
        l1Calldata: UnknownStrategy.interface.encodeFunctionData(
          "thisFunctionWillTransferFromRelayerAndExpectUnsupportedRiskLevel",
          [MockERC20.address, 0]
        ),
      });

      await snapshotGasCost(tx).should.emit(
        UnknownStrategy,
        "TransferFromRelayerFailedWithUnsupportedRiskLevel"
      );
    });

    it("does not allow calling the escrow directly", async function () {
      const L1_NovaApprovalEscrow = (
        await ethers.getContractFactory("L1_NovaApprovalEscrow")
      ).attach(await L1_NovaExecutionManager.L1_NOVA_APPROVAL_ESCROW());

      await L1_NovaApprovalEscrow.transferApprovedToken(
        ethers.constants.AddressZero,
        0,
        ethers.constants.AddressZero,
        ethers.constants.AddressZero
      ).should.be.revertedWith("UNAUTHORIZED");
    });
  });

  describe("hardRevert", function () {
    it("should revert with the proper message", async function () {
      await L1_NovaExecutionManager.hardRevert().should.be.revertedWith(
        await L1_NovaExecutionManager.HARD_REVERT_TEXT()
      );
    });
  });
});
