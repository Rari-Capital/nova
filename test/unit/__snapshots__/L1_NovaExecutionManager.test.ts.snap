// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`L1_NovaExecutionManager exec/execWithRecipient should properly execute a minimal exec 1`] = `143365`;

exports[`L1_NovaExecutionManager exec/execWithRecipient should properly execute a stateful exec 1`] = `163341`;

exports[`L1_NovaExecutionManager exec/execWithRecipient soft revert should not exec to revert 1`] = `158499`;

exports[`L1_NovaExecutionManager transferFromRelayer should transfer an arbitrary token to a strategy when requested 1`] = `188090`;
